text/0000-localizing-api-messages.md

```diff
@@ -0,0 +1,135 @@
+
+- Start Date: 03-29-2022
+- RFC PR: 
+- FOLIO Issue: 
+- Submitter : Radhakrishnan Gopalakrishnan (rgopalakrishnan@ebsco.com)
+- Co-Submitter(s) : Zak Burke (zburke@ebsco.com)
+- Sub Group :
+
+# Localizing API (Backend) Messages
+
+## Summary
+  The purpose of this RFC is to define an approach that allows API developer to provide
+messages in the API response depending on the language preference set in the request. 
+
+
+## Motivation
+
+- Help users understand the information coming from the application backend using a language they understand better
+- To allow FOLIO platform to be used by users across the world 
+- To create a tremendous opportunity for growth that we can never could have achieved within just one country.
+
+### Terminology
+- Static Message - E.g. _This is something you need translate_
+- Static Message with placeholder - _You can ONLY renew {n} times_
+ 
+
+### In Scope Requirements/Use cases
+- Return localized messages based on the value passed in the accept-language header
+- Handle static messages, static messages with placeholders
+
+### Out of Scope Requirements/Use cases
+- [Plural Syntax](https://wiki.folio.org/display/I18N/How+To+translate+FOLIO#HowTotranslateFOLIO-Pluralsyntax)
+- Naming convention to be used for keys
+- Process for managing (decoupled from lokalise.com) translations
+- Returning formatted (HTML/Markdown) messages
+- Usage of [soft hyphen](https://wiki.folio.org/display/I18N/How+To+translate+FOLIO#HowTotranslateFOLIO-Softhyphentobreakwords) to break messages
+- Support customization per tenant
+- Process for back porting API/Backend messages similar to what we have for the front end 
+- Support for controlled vocabulary (runtime data/data coming from DB tables. E.g. Patron Groups for tenants)
+
+## Detailed Explanation/Design
+- Locale specific translations MUST be stored in a properties file
```

Comment:
Are property files specific to Java?

Reply:
Every framework/language has a way of loading config info using a generic key value pair format. So, I wouldn't say it is java specific. That said, along with loading config info using this format, in java, language specific resource bundles are specified in this format to load language translations.

Reply:
> Every framework/language has a way of loading config info using a generic key value pair format. So, I wouldn't say it is java specific. 

I would imagine that each of those languages will do that slightly differently, even if it's only in naming, location and minor format changes e.g. a version header.

> That said, along with loading config info using this format, in java, language specific resource bundles are specified in this format to load language translations.

I think it is important that we separate the general protocol decision from the language specific implementation guidance.

FOLIO has a history of mixing the two together which has led to folks misunderstanding what it is they are expected to do.

Reply:
The format is as simple as 

a=b
c=d

The file name could be anything.

So, instead of using the properties file, I could refer them as key value pairs and use .properties as an example for Java. Will that be better?

Reply:
> So, instead of using the properties file, I could refer them as key value pairs and use .properties as an example for Java. Will that be better?

Yes, I think splitting the Java specific standard (because it says MUST) from the parts that are intended to be general is a good idea.

Personally (and based upon the TC conversation, other folks don't support this) I consider these files to be part of an interface between the module source and the translations process.

A part of the reason why I'm concerned about this aspect is that I think the decision is based upon it being convenient for Java modules. Much like I imagine the choice of JSON for the UI (JS based) modules was convenient for them.

Assuming this goes through, we will have two different file formats for translations. Meaning whatever part of the system / operations / infrastructure / community need to interact with them have to accommodate both of those formats.

It wouldn't surprise me if the next time we accept a new language / runtime as common into FOLIO, we'll end up with a third format because it was convenient for that environment too.

We already have back end modules that use other tools (node.js, Perl, grails) and I imagine we are expecting those modules to comply with the standards outlined here as well.